File: src/main/scala/s/decorator/delegate/FullBorderDelegate.scala
package s.decorator.delegate

final class FullBorderDelegate(private val display: Display) extends DisplayDelegate {
  override def columns: Int = 1 + display.columns + 1

  override def rows: Int = 1 + display.rows + 1

  override def rowText(row: Int): String =
    if (row == 0) "+" + makeLine('-', display.columns) + "+"
    else if (row == display.rows + 1) "+" + makeLine('-', display.columns) + "+"
    else "|" + display.rowText(row - 1) + "|"

  private def makeLine(ch: Char, count: Int) = String.valueOf(ch).repeat(Math.max(0, count))
}

File: src/main/java/j/decorator/delegate/FullBorderDelegate.java
package j.decorator.delegate;

public final class FullBorderDelegate implements DisplayDelegate {
  private final Display display;

  public FullBorderDelegate(Display display) {
    this.display = display;
  }

  @Override
  public int getColumns() {
    return 1 + display.getColumns() + 1;
  }

  @Override
  public int getRows() {
    return 1 + display.getRows() + 1;
  }

  @Override
  public String getRowText(int row) {
    if (row == 0) {
      return "+" + makeLine('-', display.getColumns()) + "+";
    } else if (row == display.getRows() + 1) {
      return "+" + makeLine('-', display.getColumns()) + "+";
    } else {
      return "|" + display.getRowText(row - 1) + "|";
    }
  }

  private String makeLine(char ch, int count) {
    return String.valueOf(ch).repeat(Math.max(0, count));
  }
}
